cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,156 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.operation;
+
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.operation.LongRunningOperation;
+import io.cdap.cdap.common.operation.LongRunningOperationRequest;
+import io.cdap.cdap.proto.id.OperationRunId;
+import io.cdap.cdap.proto.operationrun.OperationError;
+import io.cdap.cdap.proto.operationrun.OperationType;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link OperationRunner} to run an operation in the same service.
+ */
+public class InMemoryOperationRunner<T> extends AbstractOperationRunner<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRunner.class);
+
+  private final OperationStatePublisher statePublisher;
+  private final Lock lock;
+  private OperationDriver<T> driver;
+
+  /**
+   * Default constructor.
+   *
+   * @param operations A map of {@link OperationType} to {@link LongRunningOperation} injected
+   *     via {@link com.google.inject.multibindings.MapBinder}
+   * @param statePublisher Publishes the current operation state.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
+      OperationStatePublisher statePublisher) {
+    super(operations);
+    this.statePublisher = statePublisher;
+    this.lock = new ReentrantLock();
+  }
+
+  // TODO(samik) It is currently rudimentary implementation to unblock the state management
+  //  Handle states, publish failures and heartbeat messages
+  @Override
+  public void run(LongRunningOperationRequest<T> request)
+      throws OperationTypeNotSupportedException, InvalidRequestTypeException {
+    lock.lock();
+    try {
+      if (driver != null) {
+        throw new IllegalStateException("Operation already running");
+      }
+      driver = new OperationDriver<>(getOperation(request), request, statePublisher);
+      driver.start();
+      statePublisher.publishRunning();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  // TODO(samik): Implement graceful shutdown
+  @Override
+  public void stop(long timeout, TimeUnit timeoutUnit) throws Exception {
+    lock.lock();
+    try {
+      if (driver == null) {
+        throw new IllegalStateException("No operation is running");
```

Comment:
I think it's better not to throw but rather just return. This makes the caller code nicer.